tests/parametric/test_dynamic_configuration.py

```diff
@@ -0,0 +1,349 @@
+
+import pytest
+
+
+parametrize = pytest.mark.parametrize
+
+
+DEFAULT_SAMPLE_RATE = 1.0
+TEST_SERVICE = "test_service"
+TEST_ENV = "test_env"
+DEFAULT_ENVVARS = {
+    "DD_SERVICE": TEST_SERVICE,
+    "DD_ENV": TEST_ENV,
+    # Needed for .NET until Telemetry V2 is released
+    "DD_INTERNAL_TELEMETRY_V2_ENABLED": "1",
+    # Decrease the heartbeat/poll intervals to speed up the tests
+    "DD_TELEMETRY_HEARTBEAT_INTERVAL": "0.2",
+    "DD_REMOTE_CONFIG_POLL_INTERVAL_SECONDS": "0.2",
+}
+
+
+def send_and_wait_trace(test_library, test_agent, **span_kwargs) -> List[Span]:
+    with test_library.start_span(**span_kwargs):
+        pass
+    traces = test_agent.wait_for_num_traces(num=1, clear=True)
+    assert len(traces) == 1
+    return traces[0]
+
+
+def _default_config(service: str, env: str) -> Dict[str, Any]:
+    return {
+        "action": "enable",
+        "service_target": {"service": service, "env": env},
+        "lib_config": {
+            # v1 dynamic config
+            "tracing_sampling_rate": None,
+            "log_injection_enabled": None,
+            "tracing_header_tags": None,
+            # v2 dynamic config
+            "runtime_metrics_enabled": None,
+            "tracing_debug": None,
+            "tracing_service_mapping": None,
+            "tracing_sampling_rules": None,
+            "span_sampling_rules": None,
+            "data_streams_enabled": None,
+        },
+    }
+
+
+def _set_rc(test_agent, config: Dict[str, Any]) -> None:
+    cfg_id = hash(json.dumps(config))
+    test_agent.set_remote_config(
+        path="datadog/2/APM_TRACING/%s/config" % cfg_id, payload=config,
+    )
+
+
+def set_and_wait_rc(test_agent, config_overrides: Dict[str, Any]) -> Dict:
+    """Helper to create an RC configuration with the given settings and wait for it to be applied.
+
+    It is assumed that the configuration is successfully applied.
+    """
+    rc_config = _default_config(TEST_SERVICE, TEST_ENV)
+    for k, v in config_overrides.items():
+        rc_config["lib_config"][k] = v
+
+    _set_rc(test_agent, rc_config)
+
+    # Wait for both the telemetry event and the RC apply status.
+    test_agent.wait_for_telemetry_event("app-client-configuration-change", clear=True)
+    return test_agent.wait_for_rc_apply_state("APM_TRACING", state=2, clear=True)
+
+
+def assert_sampling_rate(trace: List[Dict], rate: float):
+    """Asserts that a trace returned from the test agent is consistent with the given sample rate.
+
+    This function assumes that all traces are sent to the agent regardless of sample rate.
+
+    For the day when that is not the case, we likely need to generate enough traces to
+        1) Validate the sample rate is effective (approx sample_rate% of traces should be received)
+        2) The `_dd.rule_psr` metric is set to the correct value.
+    """
+    # This tag should be set on the first span in a chunk (first span in the list of spans sent to the agent).
```

Comment:
Tag `_dd.rule_psr` is added to the _root_ span, which is not guaranteed to be the first span in a payload (the span order is not defined).